stl/src/atomic.cpp

```diff
@@ -4,33 +4,87 @@
 // implement shared_ptr spin lock
 
 #include <yvals.h>
-
+#include <atomic>
 #include <intrin.h>
-#pragma warning(disable : 4793)
+#include <libloaderapi.h>
+#include <synchapi.h>
 
-_EXTERN_C
+#pragma warning(disable : 4793) // '%s' : function is compiled as native code
 
-// SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
-volatile long _Shared_ptr_flag;
+namespace {
+    // MUTEX FOR shared_ptr ATOMIC OPERATIONS
+    SRWLOCK _Shared_ptr_lock = SRWLOCK_INIT;
 
-_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
-#ifdef _M_ARM
-    while (_InterlockedExchange_acq(&_Shared_ptr_flag, 1)) {
-        __yield();
+#if _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA
+    // SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
+    long _Shared_ptr_flag;
+
+
+    struct _Srw_functions_table {
+        _STD atomic<decltype(&::AcquireSRWLockExclusive)> _Pfn_AcquireSRWLockExclusive{nullptr};
+        _STD atomic<decltype(&::ReleaseSRWLockExclusive)> _Pfn_ReleaseSRWLockExclusive{nullptr};
+    };
+
+    _Srw_functions_table _Table;
+
+    bool _Acuire_srw_functions() {
+        enum class _Shared_ptr_api_level : char {
+            __has_nothing = false,
+            __has_srwlock = true,
+            __not_set,
+        };

```

Comment:
It's unusual (in our codebase) to convert `false` and `true` to `char` like this. I would expect these to be explicitly given the values `0`, `1`, and `2`.

Reply:
avoided `bool`, now it is really just an enum